### ISSUE TITLE
Allow empty string as delimiter for `split`

### ISSUE THREAD

**Issue by cosinequanon** [prestodb/presto#5290]
Currently you cannot provide an empty string as a delimiter for the `split` function. The error is

```
> select split('abc', '');
Query 20160517_035643_07058_prqtw failed: The delimiter may not be the empty string
```

The expected behavior is that each character in the string will be it's own value in the array, so

```
> select split('abc', '')
   _col0
-----------
 [a, b, c]
(1 row)
```

Which would be equivalent to `ARRAY['a', 'b', 'c']`

**Comment by martint** [comment]
Interestingly, `regex_split` returns `['', 'a', 'b', 'c', '']`. We should probably fix that one, too, and make them consistent.

**Comment by MariaPetropoulou** [comment]
Hi! I am an undergaduate student and I have to solve an issue for a group project of mine. I saw the requirements above and I would like to try to solve this issue. Please inform me as soon as possible.

**Comment by martint** [comment]
@MariaPetropoulou, go for it!

**Comment by MariaPetropoulou** [comment]
Thanks! Is there any tip you could give me concernig this issue?

**Comment by martint** [comment]
Here's the implementation of that function: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java#L280

**Comment by MariaPetropoulou** [comment]
Thank you very much!

**Comment by MariaPetropoulou** [comment]
Can you check this? 

 public static Block split(@SqlType("varchar(x)") Slice string, @SqlType(StandardTypes.VARCHAR) Slice delimiter, @SqlType(StandardTypes.BIGINT) long limit)
    {
        checkCondition(limit > 0, INVALID_FUNCTION_ARGUMENT, "Limit must be positive");
        checkCondition(limit <= Integer.MAX_VALUE, INVALID_FUNCTION_ARGUMENT, "Limit is too large");
        
        //We removed the CheckCondition about delimeter being an empty string
        
        BlockBuilder parts = VARCHAR.createBlockBuilder(new BlockBuilderStatus(), 1, string.length());
        // If limit is one, the last and only element is the complete string
        if (limit == 1) {
            VARCHAR.writeSlice(parts, string);
            return parts.build();
        }

        int index = 0;
        if (delimeter.length()==0){ 
        //When the delimeter is an empty string you dont need the splitIndex fuction.
            while (index < string.length()) {
            
         
          VARCHAR.writeSlice(parts, string, index, index);
          index++;
        }
          
        } else if (delimeter.length()==0) {
        while (index < string.length()) {
            int splitIndex = string.indexOf(delimiter, index);
            // Found split?
            if (splitIndex < 0) {
                break;
            }
            // Add the part from current index to found split
            VARCHAR.writeSlice(parts, string, index, splitIndex - index);
            // Continue searching after delimiter
            index = splitIndex + delimiter.length();
            // Reached limit-1 parts so we can stop
            if (parts.getPositionCount() == limit - 1) {
                break;
            }
        }
                
        // Rest of string
        VARCHAR.writeSlice(parts, string, index, string.length() - index);
                }
        return parts.build();
    }

**Comment by ebd2** [comment]
The proper way to do this would be by opening a Pull Request. Github has tools that let people comment on specific lines of your code.

Before you do that though, have you built the code with maven? This doesn't look like it conforms to Presto's checkstyle guidelines.

Once you have cloned the repository and applied your change, you can build the code by running `./mvnw clean install` from the root of your repository. That will take a while, especially the first time because maven needs to download a bunch of dependencies. You can speed up the build by including `-DskipTests` at the end of that command.

**Comment by ebd2** [comment]
Building the code with maven is covered on the Presto project's README.md, by the way. If you find that you have bugs, you can run Presto under your IDE (typically Intellij) and debug it following the instructions on the README too.

**Comment by MariaPetropoulou** [comment]
Could you please tell me where is the implementation of regex_split(function)??
Ok, thanks for your help, I will build the code with maven 
Thank you very much and sorry for the trouble.

**Comment by philippu** [comment]
It's been a few months. @MariaPetropoulou are you still working on this? Otherwise, I'm happy to take over.

**Comment by yuananf** [comment]
@philippu there is a PR about this issue#5406 , not sure whether it works

**Comment by philippu** [comment]
Thanks. I will start from Maria's PR.

**Comment by philippu** [comment]
I was experimenting with empty strings and noticed some surprising semantics. Splitting the empty string returns an array with a single element, the empty string. So for example,

`SPLIT('', '.')` returns `['']`

I would have expected an empty array. Are the existing semantics intentional?

Also, I'd love to know what `SPLIT('', '')` is supposed to return. Since the semantics of splitting by an empty delimiter are presumably "return an array where every character is an element", I would expect an empty array. But if we retain the semantics of returning `['']` for `SPLIT('', '.')`, then returning an empty array for `SPLIT('', '')` becomes very surprising.

Alternatively, we may choose to return `['', 'a', 'b', 'c', '']` for `SPLIT('abc', '')`, which looks confusing on first sight, but is consistent with most regex libraries I am aware of. It is also what REGEX_SPLIT returns. The semantics look wacky on first sight, but are clear, when considering '' to match the "empty space between characters, including the space before and after the string".

**Comment by philippu** [comment]
Oh, and to make things _really_ interesting, `REGEXP_SPLIT('', '')` returns `['', '']`.

**Comment by electrum** [comment]
I think returning an empty string is the right behavior for `split('',, 'x')`. This is the same as splitting a non-empty string on a non-matching delimiter, i.e. `split('abc', 'x')` which would return `'abc'`.

I don't think that `split` and `regex_split` need to be consistent. Matching on a regex is a different operation since a regex can match things that aren't part of the string. For example, matching on a word boundary or the start of a string. What does it mean if you split using`^`?

The current behavior of requiring a non-empty delimiter is reasonable and avoids confusing about the many possible interpretations.

Thus, I think we should keep the current behavior.

**Comment by electrum** [comment]
The behavior of regex to return empty first and last elements makes sense given that a regex can match boundaries. For example, the string 'abc', with word boundaries `*`, is `*abc*`. Thus splitting on a word boundary would result in `['', 'abc', '']`. An empty string has the same behavior if you consider an empty string to be a boundary that exists between each character.

**Comment by electrum** [comment]
TLDR: split is different becauses it matches strings (which have a minimum length of one), not arbitrary boundaries (which can be zero length).

**Comment by martint** [comment]
> I don't think that split and regex_split need to be consistent. Matching on a regex is a different operation since a regex can match things that aren't part of the string. For example, matching on a word boundary or the start of a string. 

I'm not sure I agree. `split` could certainly be modeled as a special case of `regex_split`, where split is just syntactic sugar for using regexp_split with a "literal" string. 

E.g., `split(string, 'abc')` would be equivalent to `regex_split(string, /\Qabc\E/)`. Therefore, `split(string, '')` would be the same as `regex_split(string, //)`. 

It think that interpretation is simple and elegant, and doesn't require users to be aware of two different models.

**Comment by electrum** [comment]
> split could certainly be modeled as a special case of regex_split, where split is just syntactic sugar for using regexp_split with a "literal" string.

That's reasonable, but I don't think it helps the average user to have an intuition about what splitting an empty string might mean, nor do I think there is a good use case for allowing it.

**Comment by philippu** [comment]
There is a pretty good use case: split the string into individual characters.

A different issue @electrum brought up in #5867 is that SPLIT(..., '') becomes inconsistent with SPLIT_PART when returning empty string elements at the beginning and end of the array, because SPLIT_PART does not count these towards the index parameter.

Being consistent with SPLIT_PART feels more important than being consistent with REGEX_SPLIT, so from that perspective the better option is to use the original semantics proposed by @cosinequanon and not return the empty strings. That way we can still add new useful functionality without breaking existing queries.

So, what's the vote?

**Comment by martint** [comment]
Well, `split_part` is currently inconsistent with `split`, since it allows an empty delimiter string. 

Another oddity is that while `split_('abc', '', 1)` returns `'a'`, `split_part('abc', 'a', 1)` returns `''`, so it doesn't even seem consistent with itself.

**Comment by davoscollective** [comment]
@electrum Thanks, the explanation on how `regexp_split`  with a  matches boundaries was helpful. I ended up using this hideous negative lookbehind & negative lookahead to split on any character except the beginning and end of line in order to convert a string to an array of char (varchar(1)) without the empty first and last element of the output array

`regexp_split(stringcolumnname, '(?<!^)(?!$)') `

**Comment by ssaumitra** [comment]
Hello guys,
I am a new contributor to Presto and I would like to start with this ticket.
As being discussed in past comments, I checked `split`, `regexp_split` and `split_part` behaviour in current master branch.
```
presto> select split('abcd', '');
Query 20190303_172414_00000_xub7j failed: The delimiter may not be the empty string

presto> select regexp_split('abcd', '');
      _col0
------------------
 [, a, b, c, d, ]
(1 row)


presto> select split_part('abcd', '', 1),split_part('abcd', '', 2),split_part('abcd', '', 3),split_part('abcd', '', 4),split_part('abcd', '', 5);
 _col0 | _col1 | _col2 | _col3 | _col4
-------+-------+-------+-------+-------
 a     | b     | c     | d     | NULL
(1 row)

```
I think that `split_part` works correctly without any code changes and we do not need to fix them. I have made changes to split in my pull request #12420 and this is the behaviour now, if the request gets merged
```
presto> select split('abcd', '');
     _col0
----------------
 [a, b, c, d, ]
(1 row)
```
This looks consistent with Hive
```
hive> select split('abcd', '');
OK
["a","b","c","d",""]
```
`regexp_split` and `split` are indeed inconsistent even after integrating my changes. I tried to fix `regexp_split` and I realised that `RegExp.thresholdLength` in `io.airlift.joni` is required in order to distinguish zero-length pattern. I am planning to create another PR in https://github.com/airlift/joni to propose that enhancement. Until newer version of joni is released, I think it might be ok to live with this minor inconsistency in Presto. Meanwhile, if there is simpler way to make `split` and `regexp_split` consistent, or if any enhancement is expected in my pull request, please let me know. 
Do you think this is acceptable? Please let me know if you think something needs to be changed. I will amend my PR in that case.

**Comment by wenleix** [comment]
Resolved by #12420

**Comment by sidharth-99** [comment]
Hi!! This was working till yesterday but from today onwards it's causing an error. Is anyone else facing the same issue? I'm currently facing this issues with AWS Athena.
